### PR TITLE
Correct the logic of dataflow framework.

### LINKED ISSUE #3447: NullPointerException in backward analysis.
Get `NullPointerException` when executing backward analysis:

```
> Task :dataflow:LiveVariablePlayground.main() FAILED
Exception in thread "main" java.lang.NullPointerException
	at org.checkerframework.dataflow.analysis.BackwardAnalysisImpl.performAnalysisBlock(BackwardAnalysisImpl.java:134)
	at org.checkerframework.dataflow.analysis.BackwardAnalysisImpl.performAnalysis(BackwardAnalysisImpl.java:86)
	at org.checkerframework.dataflow.cfg.CFGVisualizeLauncher.generateDOTofCFG(CFGVisualizeLauncher.java:145)
	at org.checkerframework.dataflow.cfg.playground.LiveVariablePlayground.main(LiveVariablePlayground.java:31)

Execution failed for task ':dataflow:LiveVariablePlayground.main()'.
> Process 'command '/opt/jdk/jdk-11/bin/java'' finished with non-zero exit value 1
```

A minimal test case:

```java
public class Test {
    public void test() throws Exception {
        try {
            int[] myNumbers = {1};
            System.out.println(myNumbers[1]);
        } catch (Exception e) {}
    }
}
```

This issue will be fixed in #3414.

### PR DESCRIPTION
**Pull request by xingweitian**
BackwardAnalysis:
1. Add exception block's predecessor exception block to inputs (fix null pointer exception).
2. When running `runAnalysisFor()` with an exception block, pass a copied transfer input to `callTransferFunction()`.

ForwardAnalysis:
1. When running `runAnalysisFor()` with an exception block, pass a copied transfer input to `callTransferFunction()`. In the previous dataflow framework (before #3370), we passed a non-copied transfer input (https://github.com/typetools/checker-framework/pull/3370/files#diff-d4108228a0a5407e00c090c2e97d8b6dL401-L402). This seems a bug existing in the old dataflow framework.

Resolves #3447.

### CODE REVIEW

**Comment by wmdietl** [comment]
This can wait until after the release.

**Comment by xingweitian** [comment]
Sorry, this is closed by accident.

File: dataflow/build.gradle [hunk 1]
```diff
@@ -59,3 +59,25 @@ task liveVariableTest(dependsOn: compileTestJava, group: 'Verification') {
         }
     }
 }
+
+task issue3447Test(dependsOn: compileTestJava, group: 'Verification') {
```
Comment (xingweitian) [inline dataflow/build.gradle:63]: This is ugly... Is there a better way?

File: dataflow/build.gradle [hunk 2]
```diff
@@ -59,3 +59,25 @@ task liveVariableTest(dependsOn: compileTestJava, group: 'Verification') {
...
+    description 'Test issue 3447 test case for backward analysis.'
+    inputs.file('tests/issue3447/Test.java')
+
+    outputs.file('tests/live-variable/Out.txt')
```
Comment (wmdietl) [inline dataflow/build.gradle:67]: Should these two lines use `issue3447`?

File: dataflow/tests/issue3447/Test.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+// Test case for Issue 3447:
+// https://github.com/typetools/checker-framework/issues/3447
+
+public class Test {
```
Comment (wmdietl) [inline dataflow/tests/issue3447/Test.java:4]: I would name the file and class Issue3447.

File: dataflow/build.gradle [hunk 4]
```diff
@@ -59,3 +59,25 @@ task liveVariableTest(dependsOn: compileTestJava, group: 'Verification') {
...
+    outputs.file('tests/live-variable/Test.class')
+
+    delete('tests/issue3447/Out.txt')
```
Comment (wmdietl) [inline dataflow/build.gradle:70]: Does the test even produce an `Out.txt` file?

File: dataflow/build.gradle (hunk 2 above)
Comment (xingweitian) [reply dataflow/build.gradle]: Thanks! I have fixed that.

File: dataflow/build.gradle (hunk 4 above)
Comment (xingweitian) [reply dataflow/build.gradle:66]: It does. This test relies on [dataflow/src/test/java/livevar/LiveVariable.java](https://github.com/typetools/checker-framework/blob/master/dataflow/src/test/java/livevar/LiveVariable.java).

File: dataflow/tests/issue3447/Test.java (hunk 3 above)
Comment (xingweitian) [reply dataflow/tests/issue3447/Test.java:4]: It seems that I cannot change the file name or class name from `Test` to `Issue3447`. It is hardcoded at [LiveVariable.java#L23](https://github.com/typetools/checker-framework/blob/master/dataflow/src/test/java/livevar/LiveVariable.java#L23).

Comment (wmdietl) [reply dataflow/tests/issue3447/Test.java:4]: Ah, ok.

File: dataflow/build.gradle (hunk 4 above)
Comment (wmdietl) [reply dataflow/build.gradle:66]: You're just not comparing against a golden file, but the test suite still generates the Out.txt file. Sounds good.

File: dataflow/build.gradle (hunk 1 above)
Comment (smillst) [reply dataflow/build.gradle:63]: What is ugly?  The code duplication? There may be a better way, but let's worry about that when we have a third test case.

**Pull request merged**